### PR TITLE
Support MySQL show relay log events

### LINKED ISSUE #13134: support parsing SHOW RELAYLOG EVENTS for mysql
Hi community,

This issue can be a good first try.

## Aim
support parse mysql administration statements sql `SHOW RELAYLOG EVENTS`

## Basic Qualifications
- java
- maven
- antlr4

## Detail
We have completed the clone rule in DALStatement.g4.
But have not yet implement the visit method in `shardingsphere-sql-parser/shardingsphere-sql-parser-dialect/shardingsphere-sql-parser-mysql/src/main/java/org/apache/shardingsphere/sql/parser/mysql/visitor/statement/impl/MySQLDALStatementSQLVisitor.java`
To complete this issue, I think we should complete these steps:
- Create `MySQLShowRelaylogEventsStatement` class in this package `org.apache.shardingsphere.sql.parser.sql.dialect.statement.mysql.dal`
- Add rule  in `shardingsphere-sql-parser/shardingsphere-sql-parser-engine/src/main/java/org/apache/shardingsphere/sql/parser/core/database/visitor/SQLVisitorRule.java`
- Implement the visit method in `shardingsphere-sql-parser/shardingsphere-sql-parser-dialect/shardingsphere-sql-parser-mysql/src/main/java/org/apache/shardingsphere/sql/parser/mysql/visitor/statement/impl/MySQLDALStatementSQLVisitor.java`. 
- Finally, add test for parse clone statement in `shardingsphere-test` module
## Example FYI
- Refer to this merged pr：https://github.com/apache/shardingsphere/pull/12979/files

### PR DESCRIPTION
**Pull request by flycash**
Fixes #13134.

Changes proposed in this pull request:

- overide visitLogName, visitChannelName
- introduce ChannelNameSegment

### CODE REVIEW

File: shardingsphere-sql-parser/shardingsphere-sql-parser-dialect/shardingsphere-sql-parser-mysql/src/main/java/org/apache/shardingsphere/sql/parser/mysql/visitor/statement/impl/MySQLDALStatementSQLVisitor.java [hunk 1]
```diff
@@ -452,6 +459,44 @@ public ASTNode visitShowCreateTrigger(final ShowCreateTriggerContext ctx) {
         result.setName(((IdentifierValue) visit(ctx.triggerName())).getValue());
         return result;
     }
+
+    @Override
+    public ASTNode visitShowRelaylogEvent(final MySQLStatementParser.ShowRelaylogEventContext ctx) {
+        MySQLShowRelaylogEventsStatement result = new MySQLShowRelaylogEventsStatement();
+        ChannelNameSegment channel = (ChannelNameSegment) visit(ctx.channelName());
+        result.setChannel(channel);
+        Optional.ofNullable(ctx.logName())
```
Comment (strongduanmu) [inline shardingsphere-sql-parser/shardingsphere-sql-parser-dialect/shardingsphere-sql-parser-mysql/src/main/java/org/apache/shardingsphere/sql/parser/mysql/visitor/statement/impl/MySQLDALStatementSQLVisitor.java:468]: @flycash Can you use the new visit method to encapsulate this logic? The use of lambda expressions here does not seem very clear.

File: shardingsphere-sql-parser/shardingsphere-sql-parser-dialect/shardingsphere-sql-parser-mysql/src/main/java/org/apache/shardingsphere/sql/parser/mysql/visitor/statement/impl/MySQLDALStatementSQLVisitor.java [hunk 2]
```diff
@@ -452,6 +459,44 @@ public ASTNode visitShowCreateTrigger(final ShowCreateTriggerContext ctx) {
...
+        Optional.ofNullable(ctx.LIMIT())
+                .ifPresent(node -> result.setLimit(numbers.stream()
+                        .map(ParseTree::getText)
+                        .map(NumberLiteralValue::new)
+                        .collect(Collectors.toList())));
+        return result;
+    }
+
+    @Override
+    public ASTNode visitLogName(final MySQLStatementParser.LogNameContext ctx) {
```
Comment (strongduanmu) [inline shardingsphere-sql-parser/shardingsphere-sql-parser-dialect/shardingsphere-sql-parser-mysql/src/main/java/org/apache/shardingsphere/sql/parser/mysql/visitor/statement/impl/MySQLDALStatementSQLVisitor.java:484]: @flycash Please use static import instead.

File: shardingsphere-sql-parser/shardingsphere-sql-parser-dialect/shardingsphere-sql-parser-mysql/src/main/java/org/apache/shardingsphere/sql/parser/mysql/visitor/statement/impl/MySQLDALStatementSQLVisitor.java [hunk 3]
```diff
@@ -452,6 +459,44 @@ public ASTNode visitShowCreateTrigger(final ShowCreateTriggerContext ctx) {
...
+        return visit(ctx.stringLiterals());
+    }
+    
+    @Override
+    public ASTNode visitChannelName(final MySQLStatementParser.ChannelNameContext ctx) {
+        ChannelNameSegment result = new ChannelNameSegment();
+        result.setStartIndex(ctx.start.getStartIndex());
+        result.setStopIndex(ctx.stop.getStopIndex());
+        result.setIdentifiers(ctx.identifier().stream()
```
Comment (strongduanmu) [inline shardingsphere-sql-parser/shardingsphere-sql-parser-dialect/shardingsphere-sql-parser-mysql/src/main/java/org/apache/shardingsphere/sql/parser/mysql/visitor/statement/impl/MySQLDALStatementSQLVisitor.java:493]: @flycash Please use the new visit method to encapsulate these logic.

File: shardingsphere-sql-parser/shardingsphere-sql-parser-statement/src/main/java/org/apache/shardingsphere/sql/parser/sql/dialect/statement/mysql/dal/MySQLShowRelaylogEventsStatement.java [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+@Setter
+public final class MySQLShowRelaylogEventsStatement extends AbstractSQLStatement implements DALStatement, MySQLStatement {
+    
+    private StringLiteralValue logName;
+    
+    private NumberLiteralValue from;
+    
+    private List<NumberLiteralValue> limit;
+    
+    private ChannelNameSegment channel;
```
Comment (strongduanmu) [inline shardingsphere-sql-parser/shardingsphere-sql-parser-statement/src/main/java/org/apache/shardingsphere/sql/parser/sql/dialect/statement/mysql/dal/MySQLShowRelaylogEventsStatement.java:46]: @flycash Can we use String type to replace StringLiteralValue and ChannelNameSegment? They seem to be a string.

File: shardingsphere-sql-parser/shardingsphere-sql-parser-statement/src/main/java/org/apache/shardingsphere/sql/parser/sql/dialect/statement/mysql/segment/ChannelNameSegment.java [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+import lombok.Getter;
+import lombok.Setter;
+import org.apache.shardingsphere.sql.parser.sql.common.segment.SQLSegment;
+import org.apache.shardingsphere.sql.parser.sql.common.value.identifier.IdentifierValue;
+
+import java.util.List;
+
+@Getter
+@Setter
+public class ChannelNameSegment implements SQLSegment {
```
Comment (strongduanmu) [inline shardingsphere-sql-parser/shardingsphere-sql-parser-statement/src/main/java/org/apache/shardingsphere/sql/parser/sql/dialect/statement/mysql/segment/ChannelNameSegment.java:29]: @flycash Please add javadoc for ChannelNameSegment if we need this segment.

File: shardingsphere-test/shardingsphere-parser-test/src/main/java/org/apache/shardingsphere/test/sql/parser/parameterized/asserts/statement/dal/impl/ShowRelaylogEventsStatementAssert.java [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+import org.apache.shardingsphere.test.sql.parser.parameterized.jaxb.cases.domain.statement.dal.ShowRelaylogEventsStatementTestCase;
+
+import java.util.List;
+import java.util.Optional;
+
+import static org.hamcrest.CoreMatchers.is;
+import static org.junit.Assert.assertThat;
+
+/**
+ * show relay log events statement assert.
```
Comment (strongduanmu) [inline shardingsphere-test/shardingsphere-parser-test/src/main/java/org/apache/shardingsphere/test/sql/parser/parameterized/asserts/statement/dal/impl/ShowRelaylogEventsStatementAssert.java:37]: @flycash Please use static import instead.

File: shardingsphere-test/shardingsphere-parser-test/src/main/java/org/apache/shardingsphere/test/sql/parser/parameterized/jaxb/cases/domain/segment/impl/channel/ExpectedChannelNameSegment.java [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+import lombok.Getter;
+import lombok.Setter;
+import org.apache.shardingsphere.test.sql.parser.parameterized.jaxb.cases.domain.segment.AbstractExpectedSQLSegment;
+
+import javax.xml.bind.annotation.XmlElement;
+import java.util.List;
+
+@Setter
+@Getter
+public class ExpectedChannelNameSegment extends AbstractExpectedSQLSegment {
```
Comment (strongduanmu) [inline shardingsphere-test/shardingsphere-parser-test/src/main/java/org/apache/shardingsphere/test/sql/parser/parameterized/jaxb/cases/domain/segment/impl/channel/ExpectedChannelNameSegment.java:29]: @flycash Please add final for ExpectedChannelNameSegment and add a blank line after this line.

File: shardingsphere-test/shardingsphere-parser-test/src/main/java/org/apache/shardingsphere/test/sql/parser/parameterized/jaxb/cases/domain/statement/dal/ShowRelaylogEventsStatementTestCase.java [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+import lombok.Getter;
+import org.apache.shardingsphere.test.sql.parser.parameterized.jaxb.cases.domain.segment.impl.channel.ExpectedChannelNameSegment;
+import org.apache.shardingsphere.test.sql.parser.parameterized.jaxb.cases.domain.statement.SQLParserTestCase;
+
+import javax.xml.bind.annotation.XmlAttribute;
+import javax.xml.bind.annotation.XmlElement;
+import java.util.List;
+
+/**
+ * show relay log events statement test case.
```
Comment (strongduanmu) [inline shardingsphere-test/shardingsphere-parser-test/src/main/java/org/apache/shardingsphere/test/sql/parser/parameterized/jaxb/cases/domain/statement/dal/ShowRelaylogEventsStatementTestCase.java:29]: @flycash The first letter of javadoc needs to be capitalized.

File: shardingsphere-sql-parser/shardingsphere-sql-parser-statement/src/main/java/org/apache/shardingsphere/sql/parser/sql/dialect/statement/mysql/dal/MySQLShowRelaylogEventsStatement.java [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+ * MySQL show relay log events statement.
+ */
+@ToString
+@Getter
+@Setter
+public final class MySQLShowRelaylogEventsStatement extends AbstractSQLStatement implements DALStatement, MySQLStatement {
+    
+    private StringLiteralValue logName;
+    
+    private NumberLiteralValue from;
```
Comment (strongduanmu) [inline shardingsphere-sql-parser/shardingsphere-sql-parser-statement/src/main/java/org/apache/shardingsphere/sql/parser/sql/dialect/statement/mysql/dal/MySQLShowRelaylogEventsStatement.java:42]: What is the meaning of from?

File: shardingsphere-sql-parser/shardingsphere-sql-parser-statement/src/main/java/org/apache/shardingsphere/sql/parser/sql/dialect/statement/mysql/dal/MySQLShowRelaylogEventsStatement.java [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+    
+    private List<NumberLiteralValue> limit;
```
Comment (strongduanmu) [inline shardingsphere-sql-parser/shardingsphere-sql-parser-statement/src/main/java/org/apache/shardingsphere/sql/parser/sql/dialect/statement/mysql/dal/MySQLShowRelaylogEventsStatement.java:44]: @flycash Does the limit here have the same meaning as the limit of the paging query? If yes, please use LimitSegement.

**Review by strongduanmu — CHANGES_REQUESTED** [review]

**Comment by tuichenchuxin** [comment]
@flycash Thank you for your pr. I have a suggestion, for this statement 
```
SHOW RELAYLOG EVENTS (IN logName)? (FROM NUMBER_)? (LIMIT (NUMBER_ COMMA_)? NUMBER_)? FOR CHANNEL channelName
```
I think we may not need to visit every part. What we need to know is just which this statement is,  and where it will route to execute. So, may be `MySQLShowRelaylogEventsStatement` class with logName or channelName is enough?

**Comment by flycash** [comment]
Hi, I have resolve all issues and I am not sure if I got your points. Please help to check it @tuichenchuxin @strongduanmu 
And I rename from -> position, limit -> offset, rowCount.
image

It looks like shardingsphere did not support channel options here.

**Comment by tuichenchuxin** [comment]
@flycash Please have a look at failed ci test. And I still suggest to keep logName and channelName is enough.We don't need the field about position, offset or row count.

**Comment by flycash** [comment]
@tuichenchuxin Hi, I removed these three fields and please check it again

**Review by tuichenchuxin — APPROVED** [review]
I think it's great.

**Comment by flycash** [comment]
But I don't know why the CI failed. It seems that I didn't changes these code:
image

**Comment by flycash** [comment]
I  check other PRs' CI, all failed due to the same problem. I think some merged PRs may have some problem

**Comment by tuichenchuxin** [comment]
@flycash This ci problem has been fixed, please pull master.

**Comment by tristaZero** [comment]
Hey @flycash Welcome, thanks for your continued feedback.

**Review by strongduanmu — APPROVED** [review]
@flycash Great job.

**Pull request merged**